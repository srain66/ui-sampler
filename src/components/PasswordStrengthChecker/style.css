#password-strength-checker {
  --color-strong: #10b981;
  --color-good: #3b82f6;
  --color-fair: #f59e0b;
  --color-weak: #ef4444;
}

.wrapper {
  min-width: 240px;
  width: "100%";
  margin-top: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  font-size: 0.75rem;
  padding: 0.25rem;
}

.head {
  display: flex;
  font-weight: bold;
}

.title {
  flex: none;
  margin-left: auto;
}

.title.strong {
  color: var(--color-strong);
}
.title.good {
  color: var(--color-good);
}
.title.fair {
  color: var(--color-fair);
}
.title.weak {
  color: var(--color-weak);
}

.bar-container {
  width: 100%;
  height: 0.5rem;
  background: #eee;
  border-radius: 0.5rem;
  overflow: hidden;
}

.bar {
  height: 0.5rem;
  transition: width 0.3s;
}

.bar.strong {
  background-color: green;
  width: 90%;
}
.bar.good {
  background-color: var(--color-good);
  width: 60%;
}
.bar.fair {
  background-color: var(--color-fair);
  width: 30%;
}
.bar.weak {
  background-color: var(--color-weak);
  width: 10%;
}

.description {
  color: gray;
}
