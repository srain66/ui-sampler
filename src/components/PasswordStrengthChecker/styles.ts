import styled from "styled-components";

export const Wrapper = styled.div`
  min-width: 240px;
  width: "100%";
  margin-top: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  font-size: 0.75rem;
  padding: 0.25rem;
`;

export const Header = styled.div`
  display: flex;
  font-weight: bold;
`;

export const Status = styled.p`
  flex: none;
  margin-left: auto;
`;

export const BarContainer = styled.div`
  width: 100%;
  height: 0.5rem;
  background: #eee;
  border-radius: 0.5rem;
  overflow: hidden;
`;

export const Bar = styled.div`
  height: 0.5rem;
  transition: width 0.3s;
`;

export const Text = styled.p`
  color: gray;
`;
