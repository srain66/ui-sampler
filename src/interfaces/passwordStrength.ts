export const enum PasswordStrength {
  WEAK = "Weak",
  FAIR = "Fair",
  GOOD = "Good",
  STRONG = "Strong",
}
