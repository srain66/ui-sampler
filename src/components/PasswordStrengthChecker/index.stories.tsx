import type { Meta, StoryObj } from "@storybook/react";
import PasswordStrengthChecker from "./";
import { PasswordStrength } from "@/interfaces/passwordStrength";

const meta = {
  title: "Components/PasswordStrengthChecker",
  component: PasswordStrengthChecker,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: { strength: { controls: "type" } },
} satisfies Meta<typeof PasswordStrengthChecker>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Weak: Story = {
  args: {
    strength: PasswordStrength.WEAK,
  },
};

export const Fair: Story = {
  args: {
    strength: PasswordStrength.FAIR,
  },
};

export const Good: Story = {
  args: {
    strength: PasswordStrength.GOOD,
  },
};

export const Strong: Story = {
  args: {
    strength: PasswordStrength.STRONG,
  },
};
