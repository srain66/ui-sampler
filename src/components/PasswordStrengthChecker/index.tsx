import styled from "styled-components";
import { useMemo } from "react";
import { PasswordStrength } from "@/interfaces/passwordStrength";

interface IProps {
  strength: PasswordStrength;
}

export default function PasswordStrengthChecker({ strength }: IProps) {
  const { color, width, text } = useMemo(
    () => differentByStrength[strength],
    [strength]
  );

  return (
    <Wrapper>
      <Header>
        <p>Password Strength</p>
        <Status style={{ color }}>{strength}</Status>
      </Header>
      <BarContainer>
        <Bar style={{ backgroundColor: color, width }} />
      </BarContainer>
      <Text>{text}</Text>
    </Wrapper>
  );
}

const differentByStrength = {
  [PasswordStrength.STRONG]: {
    color: "#10B981",
    width: "90%",
    text: "Your password is STRONG!",
  },
  [PasswordStrength.GOOD]: {
    color: "#3B82F6",
    width: "60%",
    text: "Your password is easily guessable.",
  },
  [PasswordStrength.FAIR]: {
    color: "#F59E0B",
    width: "30%",
    text: "Your password is easily guessable.",
  },
  [PasswordStrength.WEAK]: {
    color: "#EF4444",
    width: "10%",
    text: "Your password is easily guessable.",
  },
};

const Wrapper = styled.div`
  min-width: 240px;
  width: "100%";
  margin-top: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  font-size: 0.75rem;
  padding: 0.25rem;
`;

const Header = styled.div`
  display: flex;
  font-weight: bold;
`;

const Status = styled.p`
  flex: none;
  margin-left: auto;
`;

const BarContainer = styled.div`
  width: 100%;
  height: 0.5rem;
  background: #eee;
  border-radius: 0.5rem;
  overflow: hidden;
`;

const Bar = styled.div`
  height: 0.5rem;
  transition: width 0.3s;
`;

const Text = styled.p`
  color: gray;
`;
