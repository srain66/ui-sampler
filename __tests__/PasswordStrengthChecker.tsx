import { render } from "@testing-library/react";
import { PasswordStrength } from "@/interfaces/passwordStrength";
import PasswordStrengthChecker from "@/components/PasswordStrengthChecker";

describe("<PasswordStrengthChecker />", () => {
  it("renders when password is strong", () => {
    const { container } = render(
      <PasswordStrengthChecker strength={PasswordStrength.STRONG} />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders when password is good", () => {
    const { container } = render(
      <PasswordStrengthChecker strength={PasswordStrength.GOOD} />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders when password is fair", () => {
    const { container } = render(
      <PasswordStrengthChecker strength={PasswordStrength.FAIR} />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders when password is weak", () => {
    const { container } = render(
      <PasswordStrengthChecker strength={PasswordStrength.WEAK} />
    );
    expect(container).toMatchSnapshot();
  });
});
