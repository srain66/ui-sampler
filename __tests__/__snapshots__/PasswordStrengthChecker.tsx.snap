// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordStrengthChecker /> renders when password is fair 1`] = `
<div>
  <div
    class="wrapper"
    id="password-strength-checker"
  >
    <div
      class="head"
    >
      <p>
        Password Strength
      </p>
      <p
        class="title fair"
      >
        Fair
      </p>
    </div>
    <div
      class="bar-container"
    >
      <div
        class="bar fair"
      />
    </div>
    <p
      class="description"
    >
      Your Password is easily guessable!
    </p>
  </div>
</div>
`;

exports[`<PasswordStrengthChecker /> renders when password is good 1`] = `
<div>
  <div
    class="wrapper"
    id="password-strength-checker"
  >
    <div
      class="head"
    >
      <p>
        Password Strength
      </p>
      <p
        class="title good"
      >
        Good
      </p>
    </div>
    <div
      class="bar-container"
    >
      <div
        class="bar good"
      />
    </div>
    <p
      class="description"
    >
      Your Password is easily guessable!
    </p>
  </div>
</div>
`;

exports[`<PasswordStrengthChecker /> renders when password is strong 1`] = `
<div>
  <div
    class="wrapper"
    id="password-strength-checker"
  >
    <div
      class="head"
    >
      <p>
        Password Strength
      </p>
      <p
        class="title strong"
      >
        Strong
      </p>
    </div>
    <div
      class="bar-container"
    >
      <div
        class="bar strong"
      />
    </div>
    <p
      class="description"
    >
      Your Password is STRONG!
    </p>
  </div>
</div>
`;

exports[`<PasswordStrengthChecker /> renders when password is weak 1`] = `
<div>
  <div
    class="wrapper"
    id="password-strength-checker"
  >
    <div
      class="head"
    >
      <p>
        Password Strength
      </p>
      <p
        class="title weak"
      >
        Weak
      </p>
    </div>
    <div
      class="bar-container"
    >
      <div
        class="bar weak"
      />
    </div>
    <p
      class="description"
    >
      Your Password is easily guessable!
    </p>
  </div>
</div>
`;
